pkgs/applications/networking/cluster/k3s/default.nix

```diff
@@ -260,32 +244,74 @@ stdenv.mkDerivation rec {
     conntrack-tools
   ];
 
-  buildInputs = [
-    k3sBin
-  ] ++ k3sRuntimeDeps;
+  buildInputs = k3sRuntimeDeps;
 
-  nativeBuildInputs = [ makeWrapper ];
+  nativeBuildInputs = [
+    makeWrapper
+    rsync
+    yq-go
+    zstd
+  ];
 
-  unpackPhase = "true";
+  # embedded in the final k3s cli
+  propagatedBuildInputs = [
+    k3sCNIPlugins
+    k3sContainerd
+    k3sServer
+    runc
+  ];
+
+  # We override most of buildPhase due to peculiarities in k3s's build.
+  # Specifically, it has a 'go generate' which runs part of the package. See
+  # this comment:
+  # https://github.com/NixOS/nixpkgs/pull/158089#discussion_r799965694
+  # So, why do we use buildGoModule at all? For the `vendorSha256` / `go mod download` stuff primarily.
+  buildPhase = ''
+    # checkPhase expects this to be defined
+    getGoDirs() {
```

Comment:
Maybe drop `getGoDirs`, `subPackages` and `postInstall` and just override the check/install phases has well?

Could even disable doCheck as there is an install check and nixos test.

Reply:
`doCheck = false` seems quite reasonable, good idea. Done.

Reply:
`subPackages` and `installPhase` are a no op now.

Untested but should work.

```diff
diff --git a/pkgs/applications/networking/cluster/k3s/default.nix b/pkgs/applications/networking/cluster/k3s/default.nix
index c2af919319d..6ce1142fe40 100644
--- a/pkgs/applications/networking/cluster/k3s/default.nix
+++ b/pkgs/applications/networking/cluster/k3s/default.nix
@@ -221,7 +221,6 @@ buildGoModule rec {
   version = k3sVersion;
 
   src = k3sRepo;
-  subPackages = [ "cmd/k3s" ];
   proxyVendor = true;
   vendorSha256 = "sha256-8Yp9csyRNSYi9wo8E8mF8cu92wG1t3l18wJ8Y4L7HEA=";
 
@@ -289,14 +288,13 @@ buildGoModule rec {
     # use ./scripts/package-cli to run 'go generate' + 'go build'
 
     ./scripts/package-cli
-    mkdir -p $out/bin
-    mv dist/artifacts/k3s $out/bin/k3s
   '';
 
   # Otherwise it depends on 'getGoDirs', which is normally set in buildPhase
   doCheck = false;
 
-  postInstall = ''
+  installPhase = ''
+    install dist/artifacts/k3s -Dt $out/bin
     wrapProgram $out/bin/k3s \
       --prefix PATH : ${lib.makeBinPath k3sRuntimeDeps} \
       --prefix PATH : "$out/bin"

```

Reply:
Those changes make sense, indeed. Done, thanks!